ipld/read.go

```diff
@@ -20,188 +18,130 @@ import (
 
 var ErrRetrieveTimeout = errors.New("retrieve data timeout")
 
+type totalShares struct {
+	data [][]byte
+	err  chan error
+}
+
 // RetrieveData asynchronously fetches block data using the minimum number
 // of requests to IPFS. It fails if one of the random samples sampled is not available.
 func RetrieveData(
-	ctx context.Context,
+	parentCtx context.Context,
 	dah *da.DataAvailabilityHeader,
 	dag ipld.NodeGetter,
 	codec rsmt2d.Codec,
 ) (*rsmt2d.ExtendedDataSquare, error) {
 	edsWidth := len(dah.RowsRoots)
-	sc := newshareCounter(ctx, uint32(edsWidth))
-	// convert the row and col roots into Cids
 	rowRoots := dah.RowsRoots
 	colRoots := dah.ColumnRoots
-	// sample 1/4 of the total extended square by sampling half of the leaves in
-	// half of the rows.
-	// Sampling is done randomly here in an effort to increase
-	// resilience when the downloading node is also serving the shares back to
-	// the network. // TODO @renaynay: this should probably eventually change to
-	// just downloading the first half.
-	for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-		for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-			rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
+	shares := &totalShares{
+		data: make([][]byte, edsWidth*edsWidth),
+		err:  make(chan error),
+	}
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+
+	go fillBlockWithData(ctx, rowRoots, dag, true, shares)
+	go fillBlockWithData(ctx, colRoots, dag, false, shares)
+
+	for i := 0; i < len(dah.RowsRoots); i++ {
+		select {
+		case err := <-shares.err:
 			if err != nil {
 				return nil, err
 			}
-
-			go sc.retrieveShare(rootCid, true, row, col, dag)
+		case <-ctx.Done():
+			return nil, ctx.Err()
 		}
```

Comment:
This will remove the need to make a custom totalShares types. Let's pls avoid spamming types if we don't have to, so we keep things allocated on the stack and not the heap.

Reply:
reworked to be able to work with errgroup.